velox/type/Conversions.h

```diff
@@ -50,100 +52,124 @@ struct TruncateLegacyCastPolicy {
 
 template <TypeKind KIND, typename = void, typename TPolicy = DefaultCastPolicy>
 struct Converter {
-  template <typename T>
-  static typename TypeTraits<KIND>::NativeType cast(T) {
-    VELOX_UNSUPPORTED(
+  using TTo = typename TypeTraits<KIND>::NativeType;
+
+  template <typename TFrom>
+  static Expected<TTo> tryCast(TFrom) {
+    static const std::string kErrorMessage = fmt::format(
         "Conversion to {} is not supported", TypeTraits<KIND>::name);
+
+    return folly::makeUnexpected(Status::UserError(kErrorMessage));
   }
 };
 
+namespace detail {
+
+template <typename T, typename F>
+Expected<T> callFollyTo(const F& v) {
+  const auto result = folly::tryTo<T>(v);
+  if (result.hasError()) {
+    return folly::makeUnexpected(Status::UserError(
+        "{}", folly::makeConversionError(result.error(), "").what()));
```

Comment:
We can leave out the message as empty (and avoid creating the exception object at all) then the performance should be much closer to using numeric error code.

Reply:
This is true, but that means we loose error details for regular CAST.